File: web/src/lib/crypto.ts
/**
 * Criptografia do cliente web — tudo acontece no browser.
 *
 * O servidor guarda bytes opacos e nunca vê a senha mestra, a chave derivada nem
 * a chave do vault. As decisões abaixo estão detalhadas em
 * `docs/planos/2026-09-22-web-seguro-e-tui-completa.md` (Fase 1) e o contrato que
 * elas cumprem, em `docs/SPEC.md` seção 3.
 *
 * ⚠️ O que ZK no browser **não** protege: o operador do servidor. É ele quem
 * serve este JavaScript. "O servidor nunca vê a senha" é verdade deste código,
 * não de qualquer código que a mesma origem venha a servir amanhã.
 */

const enc = new TextEncoder()
const dec = new TextDecoder()

/**
 * Piso de iterações, **constante no bundle**. Nunca aceite este valor do
 * servidor: com escrita no banco, `UPDATE VaultConfig SET kdfIterations = 1`
 * derrubaria o custo por palpite de 600.000 hashes para 2 — fator ~300.000x.
 *
 * 600k é o piso do OWASP para PBKDF2-SHA256, o mesmo que o Bitwarden usa.
 *
 * ⚠️ Medido em 22/09/2026 nesta máquina: 600k iterações custam **~98 ms**, não
 * "~1 s" como o plano afirmava. A defesa contra força bruta *online* é o rate
 * limit do servidor, não o custo do KDF — quem escrever a rota de login não pode
 * se apoiar nesse número.
 */
export const MIN_KDF_ITERATIONS = 600_000

/** Prefixo de versão do blob. Ver `decifrar` para por que ele existe. */
const VERSAO_BLOB = "v1"

/**
 * Rótulos de HKDF. Seguem o estilo do lado Python (`src/vault/core/kdf.py`),
 * de propósito: se a Fase 4 unificar os dois vaults, isso vira troca de
 * parâmetro em vez de segunda reescrita.
 *
 * ⛔ **Nunca reaproveite um `:v1` com significado novo — crie `:v2`.** Trocar o
 * sentido de um rótulo torna indecifrável todo dado derivado sob ele.
 */
const ROTULO_AUTH = "secure-vault:web.auth:v1"
const ROTULO_ENC = "secure-vault:web.enc:v1"
const AAD_VAULTKEY = "secure-vault:web.vaultkey:v1"

/** Texto-sentinela do `keyCheck`. Mesmo papel do `KEY_CHECK_PLAINTEXT` do Python. */
const SENTINELA = "secure-vault-web-key-check"

export class ErroDeCripto extends Error {}

// ---------------------------------------------------------------------------
// base64 — sem depender de Buffer (isto roda no browser)
// ---------------------------------------------------------------------------

function paraBase64(bytes: ArrayBuffer | Uint8Array): string {
  const u8 = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes)
  let s = ""
  // Em pedaços: `String.fromCharCode(...u8)` estoura a pilha com entrada grande.
  for (let i = 0; i < u8.length; i += 0x8000) {
    s += String.fromCharCode(...u8.subarray(i, i + 0x8000))
  }
  return btoa(s)
}

function deBase64(texto: string): Uint8Array<ArrayBuffer> {
  const bruto = atob(texto)
  const u8 = new Uint8Array(bruto.length)
  for (let i = 0; i < bruto.length; i++) u8[i] = bruto.charCodeAt(i)
  return u8
}

/**
 * Zera um buffer depois do uso.
 *
 * Honestidade sobre o alcance disto: reduz a janela de exposição, **não**
 * garante que os bytes sumiram. O motor de JavaScript pode ter feito cópias
 * (realocação de heap, GC movendo objeto) fora do nosso alcance. É a mesma
 * limitação declarada no `wipe()` do lado Python.
 */
function zerar(buffer: Uint8Array<ArrayBuffer>): void {
  buffer.fill(0)
}

// ---------------------------------------------------------------------------
// Derivação
// ---------------------------------------------------------------------------

export type ChavesDerivadas = {
  /** Vai para o servidor, que guarda só o argon2id disto. Não abre nada. */
  authValue: string
  /** Abre o `wrappedVaultKey`. **Não exportável.** */
  chaveEnvelope: CryptoKey
}

/**
 * Deriva, a partir da senha mestra, um valor de autenticação e uma chave de
 * envelope — separados por HKDF com rótulo fixo.
 *
 * ⚠️ **`deriveBits` + HKDF, e não `deriveKey`.** Parece rodeio e não é: uma
 * `CryptoKey` AES não-exportável **não pode** alimentar outro KDF, então só há
 * como ter as duas coisas (chave não-exportável *e* `authValue`) passando pelos
 * bytes crus uma vez e importando o resultado com `extractable: false`.
 *
 * ⛔ **Quem trocar `false` por `true` "porque não compila" reabre o buraco
 * inteiro:** com a chave exportável, um XSS faz `crypto.subtle.exportKey` e leva
 * a chave mestra em claro. Coberto por teste.
 */
export async function derivarChaves(
  senhaMestra: string,
  salt: string,
  iteracoes: number
): Promise<ChavesDerivadas> {
  if (!Number.isInteger(iteracoes) || iteracoes < MIN_KDF_ITERATIONS) {
    // Erro VISÍVEL, nunca degradação silenciosa: este é o ponto exato onde um
    // rebaixamento de KDF seria aceito sem ninguém notar.
    throw new ErroDeCripto(
      `Parâmetro de KDF abaixo do mínimo seguro (${iteracoes} < ${MIN_KDF_ITERATIONS}). ` +
        "O vault não será aberto. Se você não mudou isso, o banco pode ter sido adulterado."
    )
  }

  const material = await crypto.subtle.importKey(
    "raw",
    enc.encode(senhaMestra),
    "PBKDF2",
    false,
    ["deriveBits"]
  )

  const bruta = await crypto.subtle.deriveBits(
    { name: "PBKDF2", salt: enc.encode(salt), iterations: iteracoes, hash: "SHA-256" },
    material,
    256
  )

  const chaveHkdf = await crypto.subtle.importKey("raw", bruta, "HKDF", false, [
    "deriveBits",
  ])

  const expandir = (rotulo: string) =>
    crypto.subtle.deriveBits(
      {
        name: "HKDF",
        hash: "SHA-256",
        salt: new Uint8Array(0),
        info: enc.encode(rotulo),
      },
      chaveHkdf,
      256
    )

  const authBits = await expandir(ROTULO_AUTH)
  const encBits = await expandir(ROTULO_ENC)

  const chaveEnvelope = await crypto.subtle.importKey(
    "raw",
    encBits,
    "AES-GCM",
    false, // ⛔ não mexer — ver o aviso acima
    ["encrypt", "decrypt"]
  )

  const authValue = paraBase64(authBits)
  zerar(new Uint8Array(encBits))

  return { authValue, chaveEnvelope }
}

// ---------------------------------------------------------------------------
// Envelope: a chave do vault não é a chave derivada da senha
// ---------------------------------------------------------------------------

export type VaultRecemCriado = {
  chaveVault: CryptoKey
  wrappedVaultKey: string
  keyCheck: string
}

/**
 * Cria a `vaultKey` aleatória e a devolve **envelopada** sob a chave derivada.
 *
 * Por que envelope e não cifrar os registros direto com a chave da senha: sem
 * ele, trocar a senha mestra significa redecifrar e recifrar os N registros no
 * browser, **sem transação**. A aba morrer no registro 40 de 200 deixa 40 blobs
 * sob a chave nova e 160 sob a velha — e o cliente não tem como saber qual é
 * qual, porque a falha de decifragem é idêntica nos dois casos. Vault meio
 * ilegível, sem diagnóstico. Com envelope, trocar a senha reescreve **um** blob.
 */
export async function criarVault(chaveEnvelope: CryptoKey): Promise<VaultRecemCriado> {
  const bytes = crypto.getRandomValues(new Uint8Array(32))
  try {
    const chaveVault = await crypto.subtle.importKey("raw", bytes, "AES-GCM", false, [
      "encrypt",
      "decrypt",
    ])
    const wrappedVaultKey = await cifrarBytes(chaveEnvelope, bytes, AAD_VAULTKEY)
    const keyCheck = await cifrarTexto(chaveVault, SENTINELA)
    return { chaveVault, wrappedVaultKey, keyCheck }
  } finally {
    // `wrapKey` do WebCrypto exigiria a chave exportável, que é justamente o que
    // não queremos — por isso passamos pelos bytes crus e os zeramos aqui.
    zerar(bytes)
  }
}

/** Abre o `wrappedVaultKey`. Falha aqui significa senha errada ou blob adulterado. */
export async function abrirVault(
  chaveEnvelope: CryptoKey,
  wrappedVaultKey: string
): Promise<CryptoKey> {
  const bytes = await decifrarBytes(chaveEnvelope, wrappedVaultKey, AAD_VAULTKEY)
  try {
    if (bytes.length !== 32) {
      throw new ErroDeCripto("wrappedVaultKey com tamanho inesperado.")
    }
    return await crypto.subtle.importKey("raw", bytes, "AES-GCM", false, [
      "encrypt",
      "decrypt",
    ])
  } finally {
    zerar(bytes)
  }
}

/**
 * Confere que a chave abre os dados **antes** de renderizar qualquer coisa.
 *
 * Por que isto não é redundante com o `abrirVault`: sem a sentinela, chave
 * errada, rebaixamento de KDF e vault meio-migrado produzem todos a **mesma
 * tela** — "nenhum item". O desfecho realista é o dono recadastrar tudo, agora
 * sob parâmetros que o atacante escolheu.
 */
export async function conferirChave(
  chaveVault: CryptoKey,
  keyCheck: string
): Promise<boolean> {
  try {
    return (await decifrarTexto(chaveVault, keyCheck)) === SENTINELA
  } catch {
    return false
  }
}

// ---------------------------------------------------------------------------
// Blob
// ---------------------------------------------------------------------------

async function cifrarBytes(
  chave: CryptoKey,
  claro: Uint8Array<ArrayBuffer>,
  aad?: string
): Promise<string> {
  // Nonce novo a CADA cifragem, nunca reaproveitado. Reusar nonce em GCM não
  // vaza só um valor: permite recuperar a chave de autenticação. É a falha mais
  // séria possível neste modo, e por isso ela é verificada por teste.
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const parametros: AesGcmParams = { name: "AES-GCM", iv }
  if (aad) parametros.additionalData = enc.encode(aad)

  const ct = await crypto.subtle.encrypt(parametros, chave, claro)
  return `${VERSAO_BLOB}.${paraBase64(iv)}.${paraBase64(ct)}`
}

async function decifrarBytes(
  chave: CryptoKey,
  blob: string,
  aad?: string
): Promise<Uint8Array<ArrayBuffer>> {
  const partes = blob.split(".")
  if (partes.length !== 3) {
    throw new ErroDeCripto("Formato de blob irreconhecível.")
  }
  const [versao, ivB64, ctB64] = partes
  if (versao !== VERSAO_BLOB) {
    // Mensagem própria, e não "falha ao decifrar": sem marcador de versão,
    // migrar de cifra vira tentativa e erro sobre dados que ninguém lê. O lado
    // Python resolveu isso de propósito (`crypto.py`, `BLOB_VERSION`).
    throw new ErroDeCripto(
      `Versão de blob desconhecida: "${versao}". Esta build entende "${VERSAO_BLOB}". ` +
        "O dado foi escrito por uma versão mais nova do Secure Vault?"
    )
  }

  const parametros: AesGcmParams = { name: "AES-GCM", iv: deBase64(ivB64) }
  if (aad) parametros.additionalData = enc.encode(aad)

  try {
    const claro = await crypto.subtle.decrypt(parametros, chave, deBase64(ctB64))
    return new Uint8Array(claro)
  } catch {
    // Não diferencia "chave errada" de "dado adulterado" para quem observa.
    throw new ErroDeCripto(
      "Não foi possível decifrar: a chave não corresponde a este dado, " +
        "ou o registro foi alterado por fora da aplicação."
    )
  }
}

async function cifrarTexto(chave: CryptoKey, texto: string): Promise<string> {
  return cifrarBytes(chave, enc.encode(texto))
}

async function decifrarTexto(chave: CryptoKey, blob: string): Promise<string> {
  return dec.decode(await decifrarBytes(chave, blob))
}

/** Cifra o payload de uma credencial sob a chave do vault. */
export async function cifrarPayload<T>(chave: CryptoKey, payload: T): Promise<string> {
  return cifrarTexto(chave, JSON.stringify(payload))
}

/** Abre um payload de credencial. Lança `ErroDeCripto` — nunca devolve `null`. */
export async function decifrarPayload<T>(chave: CryptoKey, blob: string): Promise<T> {
  return JSON.parse(await decifrarTexto(chave, blob)) as T
}

File: src/vault/core/crypto.py
"""Criptografia simétrica dos segredos guardados no vault.

Escolha: **AES-256-GCM**, da biblioteca `cryptography`.

Por que GCM e não Fernet (que também estava na mesa):

- Fernet é AES-128-CBC + HMAC-SHA256. Funciona, mas usa só 128 bits da nossa
  chave de 256 e obriga a re-codificar a chave em base64.
- GCM é AEAD nativo: autenticação e cifragem no mesmo passo, com *associated
  data* — dá para amarrar o texto cifrado ao contexto em que ele nasceu. Um blob
  copiado da coluna `encrypted_password` para a coluna `encrypted_notes` no banco
  falha na abertura, porque o AAD não bate.

Formato do blob gravado no banco:

    [1 byte versão][12 bytes nonce][ciphertext + tag de 16 bytes]

O byte de versão existe para permitir trocar de algoritmo no futuro sem
adivinhação: um blob antigo se identifica sozinho. Sem ele, migrar de cifra vira
tentativa e erro sobre dados que ninguém consegue ler.

O nonce é sorteado por `secrets.token_bytes` a **cada** operação de cifragem —
nunca reaproveitado entre registros nem entre atualizações do mesmo registro.
Reusar nonce em GCM não vaza só um valor: permite recuperar a chave de
autenticação. É a falha mais séria possível neste modo, e por isso ela é
verificada por teste (`test_crypto.py::test_nonce_nunca_se_repete`).
"""

from __future__ import annotations

import secrets

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from vault.exceptions import DecryptionError

#: Versão do formato do blob. Incrementar ao mudar algoritmo ou layout.
BLOB_VERSION = 1

NONCE_SIZE = 12  # 96 bits: tamanho recomendado pelo NIST SP 800-38D para GCM
TAG_SIZE = 16
KEY_SIZE = 32  # AES-256

#: Rótulos de *associated data*. Amarram cada texto cifrado ao campo de origem.
#:
#: Na arquitetura Zero-Knowledge a credencial inteira vira **um** blob, cifrado
#: sob `AAD_CREDENTIAL`. `AAD_MASTER_TOTP` e `AAD_KEY_CHECK` seguem em uso pelo
#: `vault_config`.
#:
#: ⚠️ `AAD_PASSWORD`, `AAD_NOTES` e `AAD_TOTP` **não são mais usados em produção**
#: — nenhum código em `src/` os referencia desde a migration `edb62ca16834`.
#: Sobrevivem por dois motivos, e nenhum deles é compatibilidade: (1) a migration
#: **recusa** converter vault populado, então não existe dado legado alcançável
#: por este código; (2) `tests/unit/test_crypto.py` precisa de dois rótulos
#: distintos para provar que decifrar com o AAD errado falha — que é a garantia
#: central do AEAD.
#:
#: **Nunca reaproveite um `:v1` para significado novo; crie `:v2`.** Trocar o
#: sentido de um rótulo torna indecifrável todo texto cifrado sob ele.
AAD_CREDENTIAL = b"secure-vault:credential.payload:v1"
AAD_PASSWORD = b"secure-vault:credential.password:v1"
AAD_NOTES = b"secure-vault:credential.notes:v1"
AAD_TOTP = b"secure-vault:credential.totp:v1"
AAD_MASTER_TOTP = b"secure-vault:vault.totp:v1"
AAD_KEY_CHECK = b"secure-vault:vault.key_check:v1"

#: Texto conhecido cifrado na criação do vault. Serve para provar, antes de
#: tentar abrir qualquer dado real, que a chave derivada é a chave certa.
KEY_CHECK_PLAINTEXT = "secure-vault-key-check"


def _validate_key(key: bytes) -> None:
    if not isinstance(key, (bytes, bytearray)):
        raise TypeError(f"chave deve ser bytes, recebido {type(key).__name__}")
    if len(key) != KEY_SIZE:
        raise DecryptionError(
            f"Chave com tamanho inválido: {len(key)} bytes, esperado {KEY_SIZE}."
        )


def encrypt(key: bytes, plaintext: str, *, aad: bytes) -> bytes:
    """Cifra `plaintext` com `key`, devolvendo o blob versionado."""
    _validate_key(key)
    nonce = secrets.token_bytes(NONCE_SIZE)
    ciphertext = AESGCM(bytes(key)).encrypt(nonce, plaintext.encode("utf-8"), aad)
    return bytes([BLOB_VERSION]) + nonce + ciphertext


def decrypt(key: bytes, blob: bytes, *, aad: bytes) -> str:
    """Abre um blob produzido por `encrypt`.

    Levanta `DecryptionError` para qualquer falha — chave errada, dado adulterado,
    blob truncado ou versão desconhecida. A mensagem nunca diferencia "chave
    errada" de "dado corrompido" para o atacante, mas diferencia o suficiente
    para o usuário legítimo saber o que fazer.
    """
    _validate_key(key)

    if not blob:
        raise DecryptionError("Texto cifrado vazio no banco.")

    minimo = 1 + NONCE_SIZE + TAG_SIZE
    if len(blob) < minimo:
        raise DecryptionError(
            f"Texto cifrado truncado: {len(blob)} bytes, mínimo {minimo}."
        )

    version = blob[0]
    if version != BLOB_VERSION:
        raise DecryptionError(
            f"Versão de formato desconhecida: {version}. "
            f"Esta build entende a versão {BLOB_VERSION}. "
            "O banco foi escrito por uma versão mais nova do Secure Vault?"
        )

    nonce = blob[1 : 1 + NONCE_SIZE]
    ciphertext = blob[1 + NONCE_SIZE :]

    try:
        plaintext = AESGCM(bytes(key)).decrypt(nonce, ciphertext, aad)
    except InvalidTag as exc:
        raise DecryptionError(
            "Não foi possível decifrar: a chave não corresponde a este dado, "
            "ou o registro foi alterado por fora da aplicação."
        ) from exc

    return plaintext.decode("utf-8")


def encrypt_optional(key: bytes, plaintext: str | None, *, aad: bytes) -> bytes | None:
    """Como `encrypt`, mas propaga `None` (campos opcionais como notas e TOTP)."""
    if plaintext is None or plaintext == "":
        return None
    return encrypt(key, plaintext, aad=aad)


def decrypt_optional(key: bytes, blob: bytes | None, *, aad: bytes) -> str | None:
    """Como `decrypt`, mas propaga `None`."""
    if blob is None:
        return None
    return decrypt(key, blob, aad=aad)


def build_key_check(key: bytes) -> bytes:
    """Cifra o texto-sentinela; guardado no vault na criação."""
    return encrypt(key, KEY_CHECK_PLAINTEXT, aad=AAD_KEY_CHECK)


def verify_key_check(key: bytes, blob: bytes) -> bool:
    """Confere se `key` é a chave deste vault, sem tocar em dado real.

    Exige a sentinela. A primeira versão aceitava `blob=None` e devolvia `True`
    — "não há o que verificar" — e isso abria um buraco sério: bastava a coluna
    `key_check` sumir (o `downgrade` do Alembic a derruba) para que `unlock`
    voltasse a aceitar como boa qualquer chave derivada. A ausência da sentinela
    é uma decisão da chamadora, tomada com contexto, e não um `True` silencioso
    aqui dentro.
    """
    if blob is None:
        raise ValueError(
            "verify_key_check exige a sentinela; trate a ausência na chamadora."
        )
    try:
        return decrypt(key, blob, aad=AAD_KEY_CHECK) == KEY_CHECK_PLAINTEXT
    except DecryptionError:
        return False


def wipe(buffer: bytearray) -> None:
    """Sobrescreve um `bytearray` com zeros.

    Limitação honesta: em CPython isto reduz a janela de exposição, mas **não**
    garante que a chave sumiu da memória — o interpretador pode ter feito cópias
    (realocação, string intermediária) fora do nosso alcance, e a página pode ter
    ido para o swap. Proteção real contra memory dumping está declarada como fora
    de escopo no README.
    """
    for i in range(len(buffer)):
        buffer[i] = 0
